### ISSUE TITLE
why are the results non-deterministic?

### ISSUE THREAD

**Issue by samurain** [SeanNaren/deepspeech.pytorch#246]
I run the an4 dataset with default settings and getting different results each time (on the same machine). 

Why are the results exactly the same? 

Here are the outputs from the first epoch of 2 runs. It seems the loss starts out the same but by the end of the first epoch it is quite different - 1.0% difference in CER on the Val set.

On other runs I have seen differences of as much as 3% in CER on the Val set after the first epoch.

Epoch: [1][1/45]	Time 6.670 (6.670)	Data 0.127 (0.127)	Loss 130.9166 (130.9166)	
Epoch: [1][2/45]	Time 6.929 (6.799)	Data 0.026 (0.076)	Loss 117.4014 (124.1590)	
Epoch: [1][3/45]	Time 8.231 (7.276)	Data 0.033 (0.062)	Loss 105.9226 (118.0802)	
Epoch: [1][4/45]	Time 8.016 (7.461)	Data 0.015 (0.050)	Loss 91.9652 (111.5514)	
Epoch: [1][5/45]	Time 8.527 (7.674)	Data 0.030 (0.046)	Loss 78.8997 (105.0211)	
Epoch: [1][6/45]	Time 9.049 (7.904)	Data 0.044 (0.046)	Loss 69.2458 (99.0585)	
Epoch: [1][7/45]	Time 8.905 (8.047)	Data 0.030 (0.044)	Loss 73.0966 (95.3497)	
Epoch: [1][8/45]	Time 9.298 (8.203)	Data 0.031 (0.042)	Loss 72.1418 (92.4487)	
Epoch: [1][9/45]	Time 10.264 (8.432)	Data 0.035 (0.041)	Loss 54.3698 (88.2177)	
Epoch: [1][10/45]	Time 10.705 (8.659)	Data 0.050 (0.042)	Loss 56.7813 (85.0741)	
Epoch: [1][11/45]	Time 11.553 (8.922)	Data 0.033 (0.041)	Loss 51.6982 (82.0399)	
Epoch: [1][12/45]	Time 11.640 (9.149)	Data 0.047 (0.042)	Loss 56.0847 (79.8770)	
Epoch: [1][13/45]	Time 12.294 (9.391)	Data 0.034 (0.041)	Loss 76.2721 (79.5997)	
Epoch: [1][14/45]	Time 11.749 (9.559)	Data 0.039 (0.041)	Loss 66.4505 (78.6605)	
Epoch: [1][15/45]	Time 12.156 (9.732)	Data 0.047 (0.041)	Loss 71.1030 (78.1566)	
Epoch: [1][16/45]	Time 12.192 (9.886)	Data 0.036 (0.041)	Loss 68.4019 (77.5469)	
Epoch: [1][17/45]	Time 13.072 (10.073)	Data 0.035 (0.041)	Loss 61.1319 (76.5814)	
Epoch: [1][18/45]	Time 14.624 (10.326)	Data 0.042 (0.041)	Loss 60.0319 (75.6619)	
Epoch: [1][19/45]	Time 13.913 (10.515)	Data 0.041 (0.041)	Loss 70.7850 (75.4053)	
Epoch: [1][20/45]	Time 15.176 (10.748)	Data 0.044 (0.041)	Loss 60.6459 (74.6673)	
Epoch: [1][21/45]	Time 15.051 (10.953)	Data 0.051 (0.041)	Loss 53.3024 (73.6499)	
Epoch: [1][22/45]	Time 15.857 (11.176)	Data 0.045 (0.042)	Loss 83.8733 (74.1146)	
Epoch: [1][23/45]	Time 16.562 (11.410)	Data 0.046 (0.042)	Loss 74.3102 (74.1231)	
Epoch: [1][24/45]	Time 17.039 (11.645)	Data 0.052 (0.042)	Loss 58.4719 (73.4710)	
Epoch: [1][25/45]	Time 17.460 (11.877)	Data 0.051 (0.042)	Loss 68.4395 (73.2697)	
Epoch: [1][26/45]	Time 18.264 (12.123)	Data 0.039 (0.042)	Loss 65.3321 (72.9644)	
Epoch: [1][27/45]	Time 18.430 (12.356)	Data 0.054 (0.043)	Loss 71.9645 (72.9274)	
Epoch: [1][28/45]	Time 18.541 (12.577)	Data 0.066 (0.044)	Loss 66.5369 (72.6992)	
Epoch: [1][29/45]	Time 19.100 (12.802)	Data 0.053 (0.044)	Loss 71.4387 (72.6557)	
Epoch: [1][30/45]	Time 19.964 (13.041)	Data 0.054 (0.044)	Loss 67.4821 (72.4833)	
Epoch: [1][31/45]	Time 20.624 (13.286)	Data 0.057 (0.045)	Loss 68.7984 (72.3644)	
Epoch: [1][32/45]	Time 21.713 (13.549)	Data 0.081 (0.046)	Loss 79.3925 (72.5840)	
Epoch: [1][33/45]	Time 20.987 (13.774)	Data 0.059 (0.046)	Loss 76.2064 (72.6938)	
Epoch: [1][34/45]	Time 22.305 (14.025)	Data 0.060 (0.047)	Loss 72.7168 (72.6945)	
Epoch: [1][35/45]	Time 23.186 (14.287)	Data 0.062 (0.047)	Loss 67.0840 (72.5342)	
Epoch: [1][36/45]	Time 25.126 (14.588)	Data 0.070 (0.048)	Loss 75.8166 (72.6253)	
Epoch: [1][37/45]	Time 25.402 (14.880)	Data 0.057 (0.048)	Loss 80.3307 (72.8336)	
Epoch: [1][38/45]	Time 25.063 (15.148)	Data 0.075 (0.049)	Loss 88.7844 (73.2533)	
Epoch: [1][39/45]	Time 25.091 (15.403)	Data 0.070 (0.049)	Loss 70.6666 (73.1870)	
Epoch: [1][40/45]	Time 28.342 (15.727)	Data 0.068 (0.050)	Loss 70.7575 (73.1263)	
Epoch: [1][41/45]	Time 28.575 (16.040)	Data 0.109 (0.051)	Loss 72.0181 (73.0993)	
Epoch: [1][42/45]	Time 30.700 (16.389)	Data 0.080 (0.052)	Loss 69.6020 (73.0160)	
Epoch: [1][43/45]	Time 34.823 (16.818)	Data 0.047 (0.052)	Loss 77.6290 (73.1233)	
Epoch: [1][44/45]	Time 39.117 (17.325)	Data 0.080 (0.052)	Loss 70.7947 (73.0703)	
Epoch: [1][45/45]	Time 40.900 (17.849)	Data 0.134 (0.054)	Loss 85.4914 (73.3057)	
Training Summary Epoch: [1]	Average Loss 73.346	
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:56<00:00,  8.14s/it]
Validation Summary Epoch: [1]	Average WER 99.320	Average CER 72.650	

From another run:
Number of parameters: 38067968
Epoch: [1][1/45]	Time 6.689 (6.689)	Data 0.125 (0.125)	Loss 130.9166 (130.9166)	
Epoch: [1][2/45]	Time 6.671 (6.680)	Data 0.025 (0.075)	Loss 117.4014 (124.1590)	
Epoch: [1][3/45]	Time 7.500 (6.953)	Data 0.023 (0.058)	Loss 105.9226 (118.0802)	
Epoch: [1][4/45]	Time 11.398 (8.064)	Data 0.053 (0.057)	Loss 91.9654 (111.5515)	
Epoch: [1][5/45]	Time 10.645 (8.581)	Data 0.040 (0.053)	Loss 78.9008 (105.0213)	
Epoch: [1][6/45]	Time 11.470 (9.062)	Data 0.036 (0.050)	Loss 69.2528 (99.0599)	
Epoch: [1][7/45]	Time 11.000 (9.339)	Data 0.085 (0.055)	Loss 72.7473 (95.3010)	
Epoch: [1][8/45]	Time 11.806 (9.647)	Data 0.041 (0.053)	Loss 71.9441 (92.3814)	
Epoch: [1][9/45]	Time 12.226 (9.934)	Data 0.029 (0.051)	Loss 55.1761 (88.2475)	
Epoch: [1][10/45]	Time 13.487 (10.289)	Data 0.029 (0.049)	Loss 55.8521 (85.0079)	
Epoch: [1][11/45]	Time 13.783 (10.607)	Data 0.035 (0.047)	Loss 51.3243 (81.9458)	
Epoch: [1][12/45]	Time 13.620 (10.858)	Data 0.032 (0.046)	Loss 58.5842 (79.9990)	
Epoch: [1][13/45]	Time 14.608 (11.146)	Data 0.031 (0.045)	Loss 70.7618 (79.2884)	
Epoch: [1][14/45]	Time 14.373 (11.377)	Data 0.035 (0.044)	Loss 67.8817 (78.4736)	
Epoch: [1][15/45]	Time 15.836 (11.674)	Data 0.035 (0.044)	Loss 66.3504 (77.6654)	
Epoch: [1][16/45]	Time 15.103 (11.888)	Data 0.047 (0.044)	Loss 71.5037 (77.2803)	
Epoch: [1][17/45]	Time 16.770 (12.176)	Data 0.036 (0.043)	Loss 60.0434 (76.2664)	
Epoch: [1][18/45]	Time 17.609 (12.477)	Data 0.040 (0.043)	Loss 58.1955 (75.2624)	
Epoch: [1][19/45]	Time 17.197 (12.726)	Data 0.061 (0.044)	Loss 68.6174 (74.9127)	
Epoch: [1][20/45]	Time 18.174 (12.998)	Data 0.039 (0.044)	Loss 57.6382 (74.0490)	
Epoch: [1][21/45]	Time 18.312 (13.251)	Data 0.039 (0.044)	Loss 49.4772 (72.8789)	
Epoch: [1][22/45]	Time 18.796 (13.503)	Data 0.040 (0.044)	Loss 68.7162 (72.6897)	
Epoch: [1][23/45]	Time 19.008 (13.743)	Data 0.045 (0.044)	Loss 76.0606 (72.8362)	
Epoch: [1][24/45]	Time 18.908 (13.958)	Data 0.043 (0.044)	Loss 50.8864 (71.9217)	
Epoch: [1][25/45]	Time 17.475 (14.099)	Data 0.043 (0.044)	Loss 73.8663 (71.9994)	
Epoch: [1][26/45]	Time 17.288 (14.221)	Data 0.055 (0.044)	Loss 62.1437 (71.6204)	
Epoch: [1][27/45]	Time 18.153 (14.367)	Data 0.076 (0.045)	Loss 73.1753 (71.6780)	
Epoch: [1][28/45]	Time 18.658 (14.520)	Data 0.049 (0.045)	Loss 62.4369 (71.3479)	
Epoch: [1][29/45]	Time 19.497 (14.692)	Data 0.051 (0.045)	Loss 74.8049 (71.4671)	
Epoch: [1][30/45]	Time 20.722 (14.893)	Data 0.060 (0.046)	Loss 61.9160 (71.1488)	
Epoch: [1][31/45]	Time 21.121 (15.094)	Data 0.038 (0.046)	Loss 80.8724 (71.4624)	
Epoch: [1][32/45]	Time 19.561 (15.233)	Data 0.056 (0.046)	Loss 79.8104 (71.7233)	
Epoch: [1][33/45]	Time 21.095 (15.411)	Data 0.060 (0.046)	Loss 71.1722 (71.7066)	
Epoch: [1][34/45]	Time 23.307 (15.643)	Data 0.057 (0.047)	Loss 81.0172 (71.9804)	
Epoch: [1][35/45]	Time 22.893 (15.850)	Data 0.117 (0.049)	Loss 66.0822 (71.8119)	
Epoch: [1][36/45]	Time 24.192 (16.082)	Data 0.061 (0.049)	Loss 79.8796 (72.0360)	
Epoch: [1][37/45]	Time 25.058 (16.325)	Data 0.072 (0.050)	Loss 83.0025 (72.3324)	
Epoch: [1][38/45]	Time 25.268 (16.560)	Data 0.066 (0.050)	Loss 79.2159 (72.5136)	
Epoch: [1][39/45]	Time 24.696 (16.769)	Data 0.067 (0.051)	Loss 79.6891 (72.6975)	
Epoch: [1][40/45]	Time 26.585 (17.014)	Data 0.076 (0.051)	Loss 69.7184 (72.6231)	
Epoch: [1][41/45]	Time 27.249 (17.264)	Data 0.071 (0.052)	Loss 58.8193 (72.2864)	
Epoch: [1][42/45]	Time 29.743 (17.561)	Data 0.068 (0.052)	Loss 70.0385 (72.2329)	
Epoch: [1][43/45]	Time 35.427 (17.976)	Data 0.080 (0.053)	Loss 81.7204 (72.4535)	
Epoch: [1][44/45]	Time 41.289 (18.506)	Data 0.057 (0.053)	Loss 67.7831 (72.3474)	
Epoch: [1][45/45]	Time 42.383 (19.037)	Data 0.102 (0.054)	Loss 88.8367 (72.6599)	
Training Summary Epoch: [1]	Average Loss 72.714	
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:58<00:00,  8.31s/it]
Validation Summary Epoch: [1]	Average WER 100.000	Average CER 73.706

**Comment by samurain** [comment]
Ok it looks like a random number seed for numpy is necessary. After adding:

np.random.seed(123456)

results are deterministic. 

I thought default behavior for the first epoch was to follow the order of the manifest but it looks like there is still some shuffling that is done.
